app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -550,22 +556,103 @@ protected void deliverError(CommCareSetupActivity receiver,
         menu.add(0, MODE_ARCHIVE, 0, Localization.get("menu.archive")).setIcon(android.R.drawable.ic_menu_upload);
+        menu.add(0, MODE_SMS, 1, Localization.get("menu.sms")).setIcon(android.R.drawable.ic_menu_upload);
         return true;
     }
 
     @Override
     public boolean onPrepareOptionsMenu(Menu menu) {
         super.onPrepareOptionsMenu(menu);
-        
+
         return true;
     }
-    
+
+    /**
+     * Scan SMS messages for texts with profile references.
+     * @param installAutomatically install automatically if reference is found
+     */
+    public void performSMSInstall(boolean installAutomatically){
+        try {
+            String profileLink = this.scanSMSLinks();
+            if (profileLink != null) {
+                // we found a valid profile link, either start install automatically
+                // or move to READY_TO_INSTALL state
+                Log.v("install", "Performing SMS install with link : " + incomingRef);
+                incomingRef = profileLink;
+                if (installAutomatically) {
+                    startResourceInstall();
+                } else {
+                    uiState = UiState.READY_TO_INSTALL;
+                    Toast.makeText(this, Localization.get("menu.sms.ready"), Toast.LENGTH_LONG).show();
+                }
+            } else {
+                // only notify if this was manually triggered, since most people won't use this
+                if(installAutomatically) {
+                    Toast.makeText(this, Localization.get("menu.sms.not.found"), Toast.LENGTH_LONG).show();
+                }
+            }
+        } catch(SignatureException e){
+            // possibly we want to do something more severe here? could be malicious
+            e.printStackTrace();
+            Toast.makeText(this, Localization.get("menu.sms.not.found"), Toast.LENGTH_LONG).show();
+        }
+    }
+
+
+    // http://stackoverflow.com/questions/11301046/search-sms-inbox
+
+    /**
+     * Scan the SMS inbox, looking for messages that meet the expected install format,
+     * and if found and verified return the discovered install link
+     * @return the verified install link, null if none found
+     * @throws SignatureException if we discovered a valid-looking message but could not verify it
+     */
+    public String scanSMSLinks() throws SignatureException{
+        String installLink = null;
+        final Uri SMS_INBOX = Uri.parse("content://sms/inbox");
+        Cursor cursor = getContentResolver().query(SMS_INBOX, null, null, null, "date desc limit 1");
+        if(cursor.moveToFirst()) {
+            Log.v("Body", cursor.getString(cursor.getColumnIndex("body")));
+            String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
+            if(textMessageBody.contains("ccapp:") && textMessageBody.contains("signature:")) {
+                installLink = parseAndVerifySMS(textMessageBody);
+            }
+        }
+        cursor.close();

```

Comment:
You'll want to use the try/finally or managed resource pattern to ensure that `cursor.close()` gets called.